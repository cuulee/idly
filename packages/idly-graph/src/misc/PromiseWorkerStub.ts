import { GetActionTypes } from '../operations/operationsTypes';

// TOFIX WorkerGetStateActionsType only does GetState actions, want to add SetState aswell
export class PromiseWorkerStub {
  /* tslint:disable */
  public _workerCb: (message: GetActionTypes) => Promise<any>;
  registerPromiseWorker(workerCb: (message: GetActionTypes) => Promise<any>) {
    this._workerCb = workerCb;
  }
  /* tslint:enable */
  public postMessage(data: {
    readonly request: any;
    readonly type: string;
  }): Promise<any> {
    const t: GetActionTypes = JSON.parse(JSON.stringify(data));
    if (!t.type) {
      throw new Error('need type');
    }
    if (!t.request) {
      throw new Error('need request');
    }
    return this._workerCb(t);
  }
}
