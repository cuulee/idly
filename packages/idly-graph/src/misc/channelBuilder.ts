import { WorkerActionsType } from '../operations/types';

/**
 * A wrapper around `promiseWorker.postMessage` api
 * its sole purpose is to provide typings and easier testing
 * the main <-> worker calls.
 * @param promiseWorker
 */
export function channelBuilder<T extends WorkerActionsType>(promiseWorker: {
  readonly postMessage: any;
}): (type: T['type']) => (request: T['request']) => Promise<string> {
  return type => {
    return request => {
      const toSend = {
        request,
        type,
      };
      return promiseWorker.postMessage(toSend).catch((e: Error) => {
        console.log('Worker Error', e.message);
        return Promise.reject(e.message);
      });
    };
  };
}
