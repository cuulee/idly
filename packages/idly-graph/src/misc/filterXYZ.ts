import area from '@turf/area';
import bboxClip from '@turf/bbox-clip';
import bboxPolygon from '@turf/bbox-polygon';

import { BBox } from 'idly-common/lib/geo/bbox';
import { mercator } from 'idly-common/lib/geo/sphericalMercator';
import { Tile } from 'idly-common/lib/geo/tile';

export function filterXyz(xyzs: Tile[], bbox: BBox, ratio: number): Tile[] {
  const mainPoly = bboxPolygon(bbox);
  return xyzs.filter(tile => {
    const secPoly = bboxPolygon(mercator.bbox(tile.x, tile.y, tile.z));
    const inter = bboxClip(secPoly, bbox);
    return area(inter) / area(secPoly) >= ratio;
  });
}
