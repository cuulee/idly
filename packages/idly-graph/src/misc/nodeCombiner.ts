import { Feature, Point } from '@turf/helpers';
import { Node } from 'idly-common/lib/osm/structures';

export function nodeCombiner(node: Node, existingProps: {}): Feature<Point> {
  return {
    geometry: {
      coordinates: [node.loc.lon, node.loc.lat],
      type: 'Point',
    },
    id: node.id,
    properties: {
      ...existingProps,
      id: node.id,
    },
    type: 'Feature',
  };
}
