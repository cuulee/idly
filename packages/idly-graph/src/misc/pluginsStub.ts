import { EntityTable } from 'idly-common/lib/osm/structures';

export function pluginsStub(
  geometry: any = {
    'osm_basic--geometry': 'line',
  },
): Promise<any> {
  return Promise.resolve({
    workers: [
      {
        pluginName: 'dummy',
        // tslint:disable no-expression-statement
        worker: (entityTable: EntityTable) => {
          const m = new Map();
          entityTable.forEach((v, k) => {
            m.set(k, geometry);
          });
          return m;
        },
      },
    ],
  });
}
