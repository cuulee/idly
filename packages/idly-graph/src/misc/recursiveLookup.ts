import {
  Entity,
  EntityId,
  EntityTable,
  EntityType,
} from 'idly-common/lib/osm/structures';

export function recursiveLookup(
  id: EntityId | undefined,
  table: EntityTable,
): Entity[] {
  if (!id) {
    return [];
  }
  const entity = table.get(id);

  if (!entity) {
    return [];
  }
  if (entity.type === EntityType.NODE) {
    return [entity];
  }
  if (entity.type === EntityType.WAY) {
    // @NOTE overriding the (Entity | undefined)[] as Entity[] assuming
    //   table would always have that node. This may or may not be correct.
    //   please check!.
    const entities = entity.nodes.map(nodeId => table.get(nodeId)) as Entity[];
    entities.push(entity);
    return entities;
  } else {
    const entities = entity.members
      .map(member => {
        // if member is relation dont dive deeper,
        // @TODO I don't really know the schema of member
        // is it id=x or ref=x. Figure it out?
        if (!member.id) {
          return [];
        }
        if (member.type === EntityType.RELATION) {
          const relation = table.get(member.id); // TOFIX when does member not have id
          if (relation) {
            return [relation];
          } else {
            return [];
          }
        }
        return recursiveLookup(member.id, table);
      })
      .reduce((prev, curr) => prev.concat(curr), []);

    entities.push(entity);
    return entities;
  }
}
