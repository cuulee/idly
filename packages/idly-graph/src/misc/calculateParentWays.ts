import { ImMap, ImSet } from 'idly-common/lib/misc/immutable';
import {
  EntityTable,
  EntityType,
  ParentWays,
} from 'idly-common/lib/osm/structures';

export function calculateParentWays(
  entityTable: EntityTable,
  parentWays: ParentWays = ImMap(),
): ParentWays {
  return parentWays.withMutations(p => {
    entityTable.forEach(w => {
      if (!w || w.type !== EntityType.WAY) {
        return;
      }
      w.nodes.forEach(nodeId => {
        /* tslint:disable */
        p.update(nodeId, (s = ImSet()) => s.add(w.id));
        /* tslint:enable */
      });
    });
  });
}
