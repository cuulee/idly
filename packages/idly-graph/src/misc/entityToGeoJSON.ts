import { Feature } from 'idly-common/lib/osm/feature';
import {
  Entity,
  EntityTable,
  EntityType,
} from 'idly-common/lib/osm/structures';

import { nodeCombiner } from './nodeCombiner';
import { wayCombiner } from './wayCombiner';

// tslint:disable:no-expression-statement object-literal-key-quotes
export function entityToGeoJson(
  entityTable: EntityTable,
  computedProps: any,
): Array<Feature<any, any>> {
  const arr: Array<Feature<any, any>> = [];
  entityTable.forEach((entity: Entity, id) => {
    if (entity.type === EntityType.NODE) {
      arr.push(nodeCombiner(entity, computedProps.get(id)));
    } else if (entity.type === EntityType.WAY) {
      arr.push(wayCombiner(entity, entityTable, computedProps.get(id)));
    } else if (entity.type === EntityType.RELATION) {
      // @TOFIX
    }
  });

  return arr;
}
