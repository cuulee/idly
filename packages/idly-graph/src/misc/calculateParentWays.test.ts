import { ImMap, ImSet } from 'idly-common/lib/misc/immutable';
import { entityTableGen } from 'idly-common/lib/osm/entityTableGen';
import { nodeFactory } from 'idly-common/lib/osm/nodeFactory';
import { wayFactory } from 'idly-common/lib/osm/wayFactory';

import { calculateParentWays } from './calculateParentWays';

// tslint:disable:no-expression-statement
describe('calculateParentWays ', () => {
  const n1 = nodeFactory({
    id: 'n1',
  });
  const n2 = nodeFactory({
    id: 'n2',
  });
  const w1 = wayFactory({
    id: 'w1',
    nodes: ['n1', 'n2'],
  });
  entityTableGen();
  it('works', () => {
    expect(calculateParentWays(ImMap(), ImSet())).toEqual(ImMap());
  });
  it('simple parentWays generation', () => {
    expect(calculateParentWays(entityTableGen([n1, n2, w1]))).toEqual(
      ImMap({
        n1: ImSet(['w1']),
        n2: ImSet(['w1']),
      }),
    );
  });
  it('takes existing parentways and adds a way', () => {
    const existingParentWay = ImMap({
      n1: ImSet(['w1']),
      n2: ImSet(['w1']),
    });

    const w2 = wayFactory({ id: 'w2', nodes: ['n1'] });
    expect(
      calculateParentWays(
        entityTableGen([n1, n2, w1, w2]),
        undefined,
        existingParentWay,
      ),
    ).toEqual(
      ImMap({
        n1: ImSet(['w1', 'w2']),
        n2: ImSet(['w1']),
      }),
    );
  });
});
