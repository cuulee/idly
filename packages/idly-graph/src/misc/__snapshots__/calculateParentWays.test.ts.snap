// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calculateParentWays  takes existing parentways and appends a way 1`] = `Immutable.Map {}`;
